packages/clay-shared/src/Overlay.tsx

```diff
@@ -0,0 +1,172 @@
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {hideOthers, suppressOthers} from 'aria-hidden';
+import React, {useCallback, useEffect, useRef} from 'react';
+
+import {Keys} from './Keys';
+import {ClayPortal} from './Portal';
+import {useInteractOutside} from './useInteractOutside';
+
+import type {Undo} from 'aria-hidden';
+
+type Props = {
+	children: React.ReactElement;
+	isCloseOnInteractOutside?: boolean;
+	isKeyboardDismiss?: boolean;
+	isModal?: boolean;
+	isOpen: boolean;
+	menuRef: React.RefObject<HTMLElement>;
+	onClose: () => void;
+	portalRef?: React.RefObject<HTMLElement>;
+	triggerRef: React.RefObject<HTMLElement>;
+};
+
+/**
+ * Inert is a new native feature to better handle DOM arias that are not
+ * assertive to a SR or that should ignore any user interaction.
+ * https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/inert
+ */
+const hasInertSupport =
+	typeof window !== 'undefined' &&
+	// @ts-ignore
+	typeof document.createElement('div').inert === 'boolean';
+
+/**
+ * Overlay component is used for components like dialog and modal.
+ * For example, Autocomplete, DatePicker, ColorPicker, DropDown are components
+ * that can use and have the same consistent behavior.
+ */
+export function Overlay({
+	children,
+	isCloseOnInteractOutside = false,
+	isKeyboardDismiss = false,
+	isModal = false,
+	isOpen = false,
+	menuRef,
+	onClose,
+	portalRef,
+	triggerRef,
+}: Props) {
+	const unsuppressCallbackRef = useRef<Undo | null>(null);
+
+	useEvent(
+		'focus',
+		useCallback(
+			(event: FocusEvent) => {
+				if (
+					portalRef &&
+					!portalRef.current?.contains(event.target as Node) &&
+					triggerRef.current &&
+					!triggerRef.current.contains(event.target as Node)
+				) {
+					onClose();
+				}
+			},
+			[onClose]
+		),
+		isOpen,
+		true,
+		[isOpen, onClose]
+	);
+
+	useEvent(
+		'keydown',
+		useCallback(
+			(event: KeyboardEvent) => {
+				if (event.key === Keys.Esc) {
+					event.stopImmediatePropagation();
+					event.preventDefault();
+
+					if (triggerRef.current) {
+						// When inert is used to suppress user interaction with the rest of
+						// the document, to retrieve the focus in the trigger we need to
+						// first undo and then move the focus.
+						if (unsuppressCallbackRef.current) {
+							unsuppressCallbackRef.current();
+							unsuppressCallbackRef.current = null;
+						}
+
+						triggerRef.current.focus();
+					}
+
+					onClose();
+				}
+			},
+			[onClose]
+		),
+		isOpen && isKeyboardDismiss,
+		true,
+		[isOpen, onClose]
+	);
+
+	useInteractOutside({
+		isDisabled: isOpen ? !isCloseOnInteractOutside : true,
+		onInteract: (event) => {
+			event.stopPropagation();
+			event.preventDefault();
+			onClose();
+		},
+		ref: portalRef ?? menuRef,
+		triggerRef,
+	});
+
+	useEffect(() => {
+		if (menuRef.current && isOpen) {
+			if (isModal && hasInertSupport) {
+				unsuppressCallbackRef.current = suppressOthers(menuRef.current);
```

Comment:
Hey @matuzalemsteles, do we need this `hasInertSupport` check here? `suppressOthers` already [checks if inert is supported](https://github.com/theKashey/aria-hidden/blob/master/src/index.ts#L207). Wouldn't this be enough?

Additionally, wouldn't this check: `if (isModal && hasInertSupport)` result in the background being disabled for all menu types not just modals? Is that the behavior we want?

By the way, I really love that this functionality is being moved to a shared location! It's such a benefit for our ability to improve and support Clay! 😄

Reply:
Hi @ethib137, oh I didn't know about the internal implementation check... but for our case `inert` should still only be used for modal cases, where there is a Dialog and DatePicker that intend to have a focus trap, this is different for a Menu , be it Actions or other common DropDown behavior, where the focus doesn't exist inside the Menu but pressing it should move the focus to the next element, using `inert` here would break these behaviors for those cases, because it also blocks any interaction from the user on these elements, so we only enable it for cases that want the focus trap.

The use of `aria-hidden` for these Menu cases is because the content is rendered in the body and it breaks the navigation based on hierarchy when using the virtual cursor of the screen readers, because it doesn't know which element to move the focus, `aria-hidden` helps us with that because it forces us to move the focus into the menu or the virtual cursor.

We will also still need to check that `inert` is supported, as it disables any user interaction on these elements, we have some behavior like pressing esc when moving focus to the trigger doesn't work, because we need to disable `inert` first and then move, this is not necessary with the use of `aria-hidden`.

Reply:
Thanks @matuzalemsteles. That makes sense. I wasn't sure why we would want to use aria-hidden if it wasn't a modal, but now I get it. It's not a big deal but if you don't want to have to maintain your own method `hasInertSupport` the aria-hidden package also exports it's own [supportsInert](https://github.com/theKashey/aria-hidden/blob/master/src/index.ts#L197) method.

Reply:
Oh great @ethib137, I'll use this!